pupil-spa/src/app/services/connection-test/connection-test.service.ts

```diff
@@ -0,0 +1,128 @@
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { Router } from '@angular/router';
+import { APP_CONFIG } from '../config/config.service';
+import { DeviceService } from '../device/device.service';
+import { StorageService } from '../storage/storage.service';
+
+declare let AzureStorage: any;
+
+@Injectable()
+export class ConnectionTestService {
+  private readonly queueName = APP_CONFIG.testSasQueueName;
+  private timeoutSeconds = 30;
+  private fibonacciN = 10000;
+  private fibonacciIterations = 50000;
+  private processingTime = -1;
+  private connectionSpeed = -1;
+
+  constructor(private http: HttpClient,
+              private router: Router,
+              private storageService: StorageService,
+              private deviceService: DeviceService) {}
+
+  async startTest() {
+    await Promise.all([
+      this.benchmarkProcessing(),
+      this.benchmarkConnection()
+    ]);
+
+    const testResults = await this.getTestResults();
+
+    try {
+      await this.submitTest(testResults);
+      this.storageService.setItem('test_status', true);
+    } catch (e) {
+      this.storageService.setItem('test_status', false);
+    }
+
+    this.router.navigate(['/ict-survey/test-completed']);
+  }
+
+  async getTestResults(): Promise<object> {
+    return {
+      device: {
+        battery: await this.deviceService.getBatteryInformation(),
+        cpu: this.deviceService.getCpuInformation(),
+        navigator: this.deviceService.getNavigatorProperties(),
+        networkConnection: this.deviceService.getNetworkInformation(),
+        screen: this.deviceService.getScreenProperties(),
+      },
+      processingTime: this.processingTime,
+      connectionSpeed: this.connectionSpeed,
+    };
+  }
+
+  public benchmarkProcessing(): Promise<void> {
+    return new Promise(resolve => {
+      const startTime = Date.now();
+
+      for (let i = 0; i < this.fibonacciIterations; i++) {
+        this.fibonacci(this.fibonacciN);
+      }
+
+      this.processingTime = Date.now() - startTime;
+      resolve();
+    });
+  }
+
+  public benchmarkConnection(): Promise<void> {
+    return new Promise(resolve => {
+      const connectionTestFile = APP_CONFIG.connectionTestFile;
```

Comment:
This might be fit for purpose, however, I have some questions that might be best asked offline.